File: 0894. All Possible Full Binary Trees/Solution.cpp
#include <vector>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x, TreeNode* left, TreeNode* right)
        : val(x)
        , left(left)
        , right(right)
    {
    }
};

class Solution {
public:
    Solution()
        : f(11)
    {
        f[1] = { new TreeNode(0) };
        for (int i = 2; i < 11; i++)
            for (int j = 1; j < i; j++)
                for (TreeNode* left : f[j])
                    for (TreeNode* right : f[i - j])
                        f[i].push_back(new TreeNode(0, left, right));
    }

    vector<TreeNode*> allPossibleFBT(int n)
    {
        if (n % 2 == 0)
            return f[0];
        return f[(n + 1) >> 1];
    }

private:
    vector<vector<TreeNode*>> f;
};

File: 0894. All Possible Full Binary Trees/Solution.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self) -> None:
        self.f = [[] for _ in range(11)]
        self.f[1].append(TreeNode(0))
        for i in range(2, 11):
            for j in range(1, i):
                for left in self.f[j]:
                    for right in self.f[i - j]:
                        self.f[i].append(TreeNode(0, left, right))

    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        if n % 2 == 0:
            return []
        return self.f[(n + 1) // 2]
